### ISSUE TITLE
Depth cache clipping not taking different heights into context

### ISSUE THREAD

**Issue by Kandob** [crest-ocean/crest#654]
**Describe the bug**
I've created a depth cache that is positioned exactly at the same height as the ocean with sufficient max terrain height. When I press bake it creates a seemingly good outline of my island:

![image](https://user-images.githubusercontent.com/14958256/93278310-92658980-f7c4-11ea-8996-4290f32eed2b.png)

But when I use it to clip the area under the terrain it only clips the parts that are exactly at the depth cache height, that means lower and higher parts are not clipped.

![Screenshot 2020-09-16 021710](https://user-images.githubusercontent.com/14958256/93278469-0bfd7780-f7c5-11ea-8814-89b141a5bf4c.jpg)

The same happens with objects that have parts hanging over the ocean, I assume this is another issue because the depth cache camera is doing a top down render.

![Screenshot 2020-09-16 021549](https://user-images.githubusercontent.com/14958256/93278570-5ed72f00-f7c5-11ea-92c6-575d54059cd1.jpg)

I also don't understand what's going on here. The terrain is pushed down but that part is still higher than the depth cache.

![Screenshot 2020-09-16 021643](https://user-images.githubusercontent.com/14958256/93278751-cd1bf180-f7c5-11ea-8160-fc1f6b874db5.jpg)

**Versions**
2019.4.0f1 Built-in

**To Reproduce**
- Create a terrain with a shoreline and make a hole in the middle of the terrain that goes deeper than the shoreline.
- Add an object that hangs over the ocean

**Comment by huwb** [comment]
Can I ask why you are using the clipping feature in this case?

Yes your right it's a top down render. There are more advanced ways to clip which we could talk about, but they are more expensive so want to understand the intention.

Sdf is a good thought but right now it only gives distance outside Island and is driven by the depth cache as input so not sure it helps at the moment.

**Comment by Kandob** [comment]
The reason why is because I have a lot of holes in the terrain that lead underground and it seems better in terms of performance to have a single clipping area under the island instead of inputting multiple.

Isn't the baked depth cache being used for clipping? It looks correct for the waves but I'm not seeing the same clipping outline like in the bake.

**Comment by huwb** [comment]
It is using the bake for clipping. It applies a threshold to the baked value to try to ensure clipping is really issue the island. Some of that area in your screenshot seems to be falling foul of that threshold and hence not being clipped. Maybe the max height needs to be increased.

I think this may end up being a juggling act that will be difficult to make work. I would consider creating some geometry where you want clipping and apply the remove surface shader on it. You could prototype with quads, and potentially merge them into a single low poly mesh to minimize cost.

**Comment by huwb** [comment]
Cc @moosichu - this could indeed be a cool application of the SDF, if we could get it inside the islands instead of just outside. Would need some kind of extension of the current approach.

**Comment by Kandob** [comment]
I've tried increasing the max height but it has no effect. Is it possible to maybe expose the threshold value in the inspector so it can be adjusted?

As you said, I might just create a mesh and use that for clipping, it's just that I'm trying to reduce the polycount because it's already pretty high.

**Comment by daleeidd** [comment]
> Is it possible to maybe expose the threshold value in the inspector so it can be adjusted?

Not a bad idea. If you need to adjust it yourself for now, it is on line 486 in the ocean shader.

I recommend using the depth cache for depth in this case since it probably isn't possible to accomplish both as Huw mentioned. Then use the clip shader which takes a texture input (_ClipSurfaceRemoveAreaTexture_) to keep poly count low.

**Comment by Kandob** [comment]
> The clip shader which takes a texture input (ClipSurfaceRemoveAreaTexture) could be used to keep poly count low.

Yes that works perfectly! I'll use that for now until the depth cache clipping is changed. Thanks for the support.
